### PR TITLE
HTML interface module + playing cards using the new module

### PR DESCRIPTION
**Pull request by nullquery**
_Ported from Baystation12 upon request of @HoaxIT_
Original PR @ https://github.com/Baystation12/Baystation12/pull/9732

This PR introduces a new method of dealing with HTML browser windows. My intention is to replace every component in the game that uses browse() with this module.

The module is documented in the main file and there's an example in the form of playing cards which have been modified to use the new system.

There are a lot of improvements to this system over the older systems, notably that it remains easy-to-use yet refreshes parts of the screen without flicker. Unlike NanoUI you specify the parts to replace directly in the source-code, there is no need to tinker with templates.

The HTML windows opened contain the latest stable/legacy version of jQuery for ease-of-use. It's possible to send Javascript calls directly using a function or you can just define something yourself through HTML.

This includes 3 possible versions of the window. The basic template is in /datum/html_interface and consists of a simple window with a browser component. /datum/html_interface/nanotrasen is the NanoUI-lookalike with a custom titlebar. I have yet to acquire icons for the minus and close buttons but I'm sure someone else can oblige in this regard (this component is not yet in use, by the time it becomes in use I hope to have those sprites).

Finally there's /datum/html_interface/cards for the playing cards. They need a subtype because the playing cards have their own resources which need to be preloaded by the client.

Future contributions to the module will be a generic way of authenticating ID cards, so that the computers on the station can be ported to make use of the new system.

Ideally NanoUI becomes deprecated, I'm not sure what the feeling about that is but I think it's necessary to keep the system as simple as possible.

image

### CODE REVIEW

File: code/modules/html_interface/html_interface.dm [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
... lorem sed lacinia. Maecenas vel posuere sapien. In lobortis odio non tincidunt ultricies. Sed consequat molestie orci et pharetra. Suspendisse potenti. Vestibulum vitae ornare risus, nec semper arcu. Duis et interdum lacus.</p><p>Etiam urna nulla, pulvinar at est auctor, varius feugiat orci. Vestibulum efficitur maximus imperdiet. Donec vehicula, leo sit amet condimentum pulvinar, urna felis aliquet velit, bibendum placerat dui libero sed tortor. Vivamus ac diam commodo nisi facilisis lacinia. Aenean a rhoncus risus, venenatis efficitur arcu. Curabitur tincidunt nulla eget augue malesuada imperdiet. Quisque ligula purus, dictum a imperdiet eget, eleifend eu leo. Phasellus massa ipsum, molestie nec pellentesque eu, scelerisque et mi. Vivamus at libero varius, lacinia magna non, imperdiet tortor. Donec scelerisque ipsum sollicitudin justo ornare accumsan. In velit orci, lobortis eget maximus et, scelerisque ut nulla. Cras sit amet finibus sapien. Aenean metus lorem, gravida a rutrum quis, varius eu arcu. Integer ac hendrerit purus. Aliquam cursus ultricies tortor. Fusce scelerisque, arcu id pellentesque accumsan, nulla turpis tempus lectus, tincidunt blandit mi nisi non metus.</p>")
```
Comment (Aranclanos) [inline code/modules/html_interface/html_interface.dm]: WAKE ME UP

**Comment by HoaxIT** [comment]
I saw some positive feedback on the forums when the topic was made back in 2014, and since most, if not all developers tend to hate NanoUI since it lacks documentation, and proper updates, this looked like an good thing.

File: code/modules/html_interface/html_interface.dm (hunk 1 above)
Comment (nullquery) [inline code/modules/html_interface/html_interface.dm]: Apologies for the lorem ipsum garbage. I've reduced the amount of characters!

**Comment by nullquery** [comment]
**To anyone looking to develop this further:** Please contact me on Github or at the very least develop from [my own repository](https://github.com/nullstation/nullstation) and then push it to your own codebase (or I'll do it for you as you've just witnessed).

File: code/modules/client/client procs.dm [hunk 2]
```diff
@@ -264,6 +264,14 @@ var/next_external_rsc = 0
...
 //send resources to the client. It's here in its own proc so we can move it around easiliy if need be
 /client/proc/send_resources()
+
+	spawn
+		// Preload the HTML interface. This needs to be done due to BYOND bug http://www.byond.com/forum/?post=1487244
+		var/datum/html_interface/hi
+		for (var/type in typesof(/datum/html_interface))
+			hi = new type(null)
+			hi.sendResources(src)
+
```
Comment (MrPerson) [inline code/modules/client/client procs.dm]: That link gives me a 404

Comment (nullquery) [inline code/modules/client/client procs.dm]: Oh. That issue is hidden for security. It's marked as resolved, but I don't think it is yet (I need to report it again, but for that I'll need a clear test case).

The issue specifies that browse_rsc works asynchronously rather than synchronously, so you can't use it in your normal flow of operations but have to preload resources first.

i.e., if you call browse_rsc() with a css file and call browse() immediately afterward, it won't find the css file because it's still being downloaded by the client. It's not meant to do that.

Comment (Aranclanos) [inline code/modules/client/client procs.dm]: its because the code is copypasted

**Comment by optimumtact** [comment]
Does this repush the entire ui on every update or does it only send changes?

One of the advantages of nanoui is that the updates happened clientside, so you only needed to stream the data changes.

File: code/modules/html_interface/html_interface.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
+var is_loading								= false;
```
Comment (optimumtact) [inline code/modules/html_interface/html_interface.js]: Why is the spacing in this file so weird?

**Comment by optimumtact** [comment]
This looks okay, kinda iffy that there are now three different methods to deliver UI's and nobody's committed to actually replacing all the others with one.

**Comment by nullquery** [comment]
Uh, yeah I'm committed to replace this with one method.

**Comment by optimumtact** [comment]
are you going to answer my question about how it handles updates?

**Comment by Aranclanos** [comment]
on space station 13 web designers never stay for longer than a week

**Comment by Chnkr** [comment]
@nullquery

**Comment by nullquery** [comment]
Huh. Didn't receive a notification until now. Sorry!

It's up to the developer what exactly is updated. You can stream bits of content or send the entire HTML. In both cases it'll update without flicker by using BYOND's version of ajax.

**Comment by nullquery** [comment]
If there is actually positive thoughts on this feature I'll go ahead and modify the rest, the reason for delays (among personal projects) is that I was hoping to get this initial bit pushed to various code bases. Unfortunately some are more reluctant to accept it (even though it doesn't conflict with other people's stuff).

**Comment by optimumtact** [comment]
:+1:  I don't have any other concerns

**Comment by nullquery** [comment]
Hi,

I've resolved the merge conflict, but so far Travis CI isn't running?

**Comment by optimumtact** [comment]
looks like it ran to me?

**Comment by nullquery** [comment]
Yay! It ran! Can someone merge this? Thanks

**Comment by d3athrow** [comment]
>on space station 13 web designers never stay for longer than a week

No shit, he said he was going to convert more things to the browser datums on /vg/ when he added *last year*

**Comment by Iamgoofball** [comment]
web designers are the only ones who can escape the curse

**Comment by nullquery** [comment]
To come to my own defense, it's not exactly "last year", it was between December '14 - January '15' that I gave this a shot.

Here is a breakdown of why I wasn't active since then:

* I had wanted to set up a server, but it died pretty quickly because the people I was working with (unrelated to /vg/) weren't committed to the project.

* I knew I'd be back later, since if anyone here is part of the curse it's me (active for 10+ years).

* I also knew that the changes I pushed through were not critical changes that needed my attention. If @d3athrow wants to he can revert the html_interface module changes even at this point in time with _zero_ loss to gameplay, unless he started using the playing cards in which case he'd have to grab the playing card code from Bay12. But I doubt the playing cards are actively being used.

I've been considering my options and thinking about how I should proceed. I'm going to give this another shot, but I want to know I can count on PR's being accepted without a lot of drama in the way -- especially about features like these that could easily be reverted later should things not work out.

My intention as before is still to implement the html_interface module everywhere. But if it turns out I don't deliver, reverting this is quite easy or you can start utilizing it yourself.

That said if there are _any_ questions about my work feel free to ask them. I'm always available to answer questions and in that regard I haven't left at all.

**Comment by nullquery** [comment]
Regarding this PR: it no longer has a merge conflict (label can be removed) and from what I understand nobody has any objections (other than perhaps the lingering doubts if I'll continue to build on this) so can someone merge this PR?? I've got myself stuck and now I can't commit anything else without adding it to this PR.

Tonight I'm hoping to do some more work, maybe port a few computers over to the new system. I'll also look if it's feasible to port the same things over in /vg/ though admittedly it has less priority for me since we're looking into using /tg/ as our base (but I'm not going back on my word).

**Comment by RemieRichards** [comment]
happening.

**Comment by nullquery** [comment]
Thanks!
